### PR TITLE
[PackageModel] Support swift-testing installed in a toolchain

### PR DESCRIPTION
**Pull request by xedin**
Properly handle swift-testing installations in toolchain/SDK.

### Motivation:

On macOS SwiftPM should prefer swift-testing installed into a custom toolchain when used. On Windows we need special logic to discover swift-testing location.

### Modifications:

- Add special swift compiler "extra" flags to favor swift-testing installed in a toolchain.
- Inject `-I`, `-L` on Windows that point to where swift-testing is installed in SDKROOT.
- Inject a path to testing on `PATH` environment variable on Windows to make sure that the library is always discoverable.

Resolves: rdar://132828246

### CODE REVIEW

File: Sources/PackageModel/UserToolchain.swift [hunk 1]
```diff
@@ -786,6 +873,40 @@ public final class UserToolchain: Toolchain {
...
+            sdkRoot = sdkDir
+        } else if let SDKROOT = environment["SDKROOT"], let sdkDir = try? AbsolutePath(validating: SDKROOT) {
+            sdkRoot = sdkDir
+        } else {
+            return .none
+        }
+
+        // The layout of the SDK is as follows:
+        //
+        // Library/Developer/Platforms/[PLATFORM].platform/Developer/Library/<Project>-[VERSION]/...
```
Comment (MaxDesiatov) [inline Sources/PackageModel/UserToolchain.swift:893]: Do we really use `Library/Developer` directory naming on Windows? I thought this was very Darwin-specific

**Review by MaxDesiatov — APPROVED** [review]
LGTM with a minor nit about a Windows-specific layout comment

File: Sources/PackageModel/UserToolchain.swift (hunk 1 above)
Comment (xedin) [reply Sources/PackageModel/UserToolchain.swift:893]: I'd think so since this was the original comment I moved, that's how we discover XCTest today.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Windows

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@swift-ci please test Windows platform

File: Sources/PackageModel/UserToolchain.swift [hunk 2]
```diff
@@ -492,16 +540,31 @@ public final class UserToolchain: Toolchain {
...
+                        let swiftTestingInstallation: AbsolutePath =
+                            platform.appending("Developer")
+                                .appending("Library")
+                                .appending("Testing-\(info.defaults.swiftTestingVersion)")
+
+                        for flag in ["-I", "-L"] {
+                            swiftTesting += try [
+                                flag,
+                                AbsolutePath(
+                                    validating: "usr/lib/swift/windows/\(triple.archName)",
```
Comment (rintaro) [inline Sources/PackageModel/UserToolchain.swift:558]: `-I` is `usr/lib/swift/windows` not in the "arch" directory

Comment (xedin) [reply Sources/PackageModel/UserToolchain.swift]: Fixed, thank you!

**Comment by rintaro** [comment]
https://github.com/apple/swift-testing/pull/581
https://github.com/swiftlang/swift-installer-scripts/pull/317
https://github.com/swiftlang/swift/pull/75515
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/swift-testing/pull/581
https://github.com/swiftlang/swift-installer-scripts/pull/317
https://github.com/swiftlang/swift/pull/75515
@swift-ci Please test Windows

**Comment by rintaro** [comment]
https://github.com/apple/swift-testing/pull/581
https://github.com/swiftlang/swift-installer-scripts/pull/317
https://github.com/swiftlang/swift/pull/75515
@swift-ci Please test

**Comment by rintaro** [comment]
Obviously cross repo PR testing from non-swift repo doesn't support changes in `update-checkout` scheme.
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
Windows test fails as it fails to decode `Info.plist`.  Since `swiftTestingVersion` is not optional, it requires corresponding swift change.

**Comment by xedin** [comment]
@rintaro maybe it would be prudent to make it optional just in case people would try to run newer swiftpm with older toolchains?

**Comment by rintaro** [comment]
Yeah, optional sounds reasonable to me.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
I'm going to land this is as a rebase because it has changes for both macOS and Windows and we should be revertible separately if we wanted to.

**Review by rintaro — APPROVED** [review]

File: Sources/PackageModel/Toolchain.swift [hunk 3]
```diff
@@ -74,20 +74,24 @@ extension Toolchain {
...
-            return try AbsolutePath(validating: "../../lib/swift/macosx", relativeTo: resolveSymlinks(swiftCompilerPath))
+            try Self.toolchainLibDir(swiftCompilerPath: self.swiftCompilerPath).appending(
+                components: ["swift", "macosx"]
+            )
         }
     }
 
     public var toolchainLibDir: AbsolutePath {
         get throws {
             // FIXME: Not sure if it's better to base this off of Swift compiler or our own binary.
```
Comment (compnerd) [inline Sources/PackageModel/Toolchain.swift:93]:
```suggestion
```

File: Sources/PackageModel/UserToolchain.swift [hunk 4]
```diff
@@ -492,16 +540,36 @@ public final class UserToolchain: Toolchain {
...
+                                    relativeTo: swiftTestingInstallation
+                                ).pathString
+                            ]
+                        }
+
                         extraSwiftCFlags = info.defaults.extraSwiftCFlags ?? []
                     }
 
-                    return ["-sdk", sdkroot.pathString] + runtime + xctest + extraSwiftCFlags
+                    return ["-sdk", sdkroot.pathString] + runtime + xctest + swiftTesting + extraSwiftCFlags
```
Comment (compnerd) [inline Sources/PackageModel/UserToolchain.swift:572]: Can we limit to the testing framework in use by any chance?

**Pull request merged**